File: easy/ts/628.ts
// https://leetcode.com/problems/maximum-product-of-three-numbers/?source=submission-noac
// Time: O(n * log(n))
// Space: O(n)
function maximumProduct(nums: number[]): number {
  if (nums.length === 3) return nums[0] * nums[1] * nums[2];
  nums.sort((a, b) => a - b);
  let leftMax = nums[0] * nums[1] * nums[nums.length - 1];
  let rightMax =
    nums[nums.length - 1] * nums[nums.length - 2] * nums[nums.length - 3];

  return leftMax > rightMax ? leftMax : rightMax;
}

File: easy/python/628.py
# https://leetcode.com/problems/maximum-product-of-three-numbers/description/?source=submission-noac
# Time: O(n * log(n))
# Space: O(n)
class Solution:
    def maximumProduct(self, nums: List[int]) -> int:
        if len(nums) == 3:
            return nums[0] * nums[1] * nums[2]
        nums.sort()
        left_sum = nums[0] * nums[1] * nums[len(nums) - 1]
        right_sum = nums[len(nums) - 1] * nums[len(nums) - 2] * nums[len(nums) - 3]
        return left_sum if left_sum > right_sum else right_sum
